decoder/decoder.cpp

```diff
@@ -0,0 +1,67 @@
+#include <iostream>
+#include <cassert>
+#include <algorithm>
+
+#include "decoder.h"
+
+/*NOTES:
+b will be translated to beq
+nop will be translated to sll
+*/
+
+using namespace Decoder;
+
+unsigned Decoder::get_bits(bit_range range, word_type from) {
+    assert(range.second > range.first);
+    int len = range.second - range.first;
+    word_type mask = ((1 << len) - 1);
+    return (from >> range.first) & mask;
+}
+    
+Instruction Decoder::decode_word(word_type word) {
+	using namespace Commands;
+
+    Instruction parsed;
+    unsigned opcode = get_bits(op_range, word);
+    if (!opcode) {//special
+        opcode = get_bits(func_range, word);
+        parsed.type = Commands::cmd_type::SPEC_CMD;	//TODO sllv, srav
+        if (std::find(shift_cmds.begin(), shift_cmds.end(), opcode) != shift_cmds.end())
+        	parsed.imm = static_cast<int>(get_bits(shift_range, word));	//warning here
+        parsed.rd = get_bits(rd_range, word);
+        parsed.rs = get_bits(rs_range, word);
+        parsed.rt = get_bits(rt_range, word);                
+    }
+    else if (opcode == 0b010000) {//COP0
+    	opcode = get_bits(func_range, word);
+        parsed.type = Commands::COP_CMD;
+    }
+    else if (opcode == j || opcode == jal) {
+    	parsed.imm = static_cast<int>(get_bits(address_range, word) << 2);	//warning here
+    	parsed.type = Commands::CMD;
+	}
+	else {
+    	parsed.type = Commands::CMD;
+    	std::array<command_name, 7> signed_ops = {
+    		addi, b_andi, b_ori, b_xori, lw, lh, lb };
```

Comment:
I suppose, it belongs to other place. Why do you have global shift_cmds, but defined this as automatic array?